File: Updog.Application/Core/Paging/PagedResultSet.cs
using System.Collections;
using System.Collections.Generic;

namespace Updog.Application.Paging {
    /// <summary>
    /// Result set that had paging applied to it.
    /// </summary>
    /// <typeparam name="T">The resource being paged.</typeparam>
    public sealed class PagedResultSet<T> : IEnumerable<T> {
        #region Properties
        /// <summary>
        /// Pagination info such as page number and page size,
        /// along with total page count.
        /// </summary>
        public PaginationInfo Pagination { get; }

        /// <summary>
        /// The items of the result set.
        /// </summary>
        public IEnumerable<T> Items { get; }
        #endregion

        #region Constructor(s)
        /// <summary>
        /// Create a new set of paged results.
        /// </summary>
        /// <param name="items">The items of the set.</param>
        /// <param name="paginationInfo">The pagination info.</param>
        public PagedResultSet(IEnumerable<T> items, PaginationInfo paginationInfo) {
            Items = items;
            Pagination = paginationInfo;
        }

        public IEnumerator<T> GetEnumerator() {
            return Items.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator() {
            return Items.GetEnumerator();
        }
        #endregion
    }
}

File: Updog.Client/src/core/pagination/paged-result-set.ts
import { PaginationInfo } from './pagination-info';

/**
 * A result set from the backend that was paged.
 */
export class PagedResultSet<T> extends Array<T> {
    /**
     * The pagination info of the result set.
     */
    public pagination: PaginationInfo;

    /**
     * Create a new paged result set.
     * @param items The items of the result set.
     * @param pagination The pagination info.
     */
    constructor(items: T[], pagination: PaginationInfo) {
        super();
        items.forEach(i => this.push(i));
        this.pagination = pagination;
    }

    /**
     * Create a new empty pagination result set.
     */
    public static Empty<T>(): PagedResultSet<T> {
        return new PagedResultSet([], new PaginationInfo(0, 0, 0));
    }
}
